include/boost/geometry/util/promote_integral.hpp

```diff
@@ -0,0 +1,276 @@
+                        void
+                    >
+            > type;
+    };
+
+    // Meta-function to get the multiprecision signed integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_signed_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::signed_magnitude
+            >::type type;
+    };
+
+    // Meta-function to get the multiprecision unsigned integer type for
+    // the given size
+    template <unsigned int Size>
+    struct multiprecision_unsigned_integer_type
+    {
+        typedef typename multiprecision_integer_type
+            <
+                Size, boost::multiprecision::unsigned_magnitude
+            >::type type;
+    };
+
+    // Define the minimum (and maximum) bit size for the
+    // multiprecision integer needed
+    // If T is the input type and P the promoted type, then the
+    // minimum number of bits for P are (below b stands for the number
+    // of bits of T):
+    // * if T is unsigned and P is unsigned: 2 * b
+    // * if T is signed and P is signed: 2 * b - 1
+    // * if T is unsigned and P is signed: 2 * b + 1
+    typedef typename boost::mpl::if_c
+        <
+            (PromoteUnsignedToUnsigned && is_unsigned),
+            boost::integral_constant
+                <
+                    unsigned int, (2 * CHAR_BIT * sizeof(T))
+                >,
+            boost::integral_constant
+                <
+                    unsigned int,
+                    (2 * CHAR_BIT * sizeof(T) + (is_unsigned ? 1 : -1))
+                >
+        >::type multiprecision_integer_min_bit_size_type;
+#endif
+
+    // Define the minimum size (in bytes) for the promoted type
+    typedef typename boost::mpl::if_c
+        <
+            (PromoteUnsignedToUnsigned && is_unsigned),
+            boost::integral_constant<std::size_t, (2 * sizeof(T))>,
+            boost::integral_constant
+                <
+                    std::size_t, (2 * sizeof(T) + (is_unsigned ? 1 : -1))
+                >
+        >::type min_size_type;
+
+    // Define the list of signed integral types we are goind to use
+    // for promotion
+    typedef boost::mpl::list
+        <
+            short, int, long
+#if defined(BOOST_HAS_LONG_LONG)
+            , boost::long_long_type
+#endif
+#if defined(BOOST_HAS_INT128)
+            , boost::int128_type
+#endif
+#if !defined(BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER)
+            , typename multiprecision_signed_integer_type
+                <
+                    multiprecision_integer_min_bit_size_type::value
+                >::type
+#endif
+        > signed_integral_types;
+
+    // Define the list of unsigned integral types we are goind to use
+    // for promotion
+    typedef boost::mpl::list
+        <
+            unsigned short, unsigned int, unsigned long, std::size_t
+#if !defined(BOOST_GEOMETRY_NO_MULTIPRECISION_INTEGER)
```

Comment:
Thanks a lot for the link.
Support for `unsigned long long` and `uint128` has now been added.
See commits af1a3da and 27e7ad6.